// @fal-ai/serverless-client/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAppId = exports.stream = exports.storage = exports.ValidationError = exports.ApiError = exports.realtime = exports.withProxy = exports.withMiddleware = exports.subscribe = exports.run = exports.queue = exports.getConfig = exports.config = void 0;
var config_1 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_1.config; } });
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return config_1.getConfig; } });
var function_1 = require("./function");
Object.defineProperty(exports, "queue", { enumerable: true, get: function () { return function_1.queue; } });
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return function_1.run; } });
Object.defineProperty(exports, "subscribe", { enumerable: true, get: function () { return function_1.subscribe; } });
var middleware_1 = require("./middleware");
Object.defineProperty(exports, "withMiddleware", { enumerable: true, get: function () { return middleware_1.withMiddleware; } });
Object.defineProperty(exports, "withProxy", { enumerable: true, get: function () { return middleware_1.withProxy; } });
var realtime_1 = require("./realtime");
Object.defineProperty(exports, "realtime", { enumerable: true, get: function () { return realtime_1.realtimeImpl; } });
var response_1 = require("./response");
Object.defineProperty(exports, "ApiError", { enumerable: true, get: function () { return response_1.ApiError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return response_1.ValidationError; } });
var storage_1 = require("./storage");
Object.defineProperty(exports, "storage", { enumerable: true, get: function () { return storage_1.storageImpl; } });
var streaming_1 = require("./streaming");
Object.defineProperty(exports, "stream", { enumerable: true, get: function () { return streaming_1.stream; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "parseAppId", { enumerable: true, get: function () { return utils_1.parseAppId; } });
